### PR TITLE
Lazy create EndpointRouting statemachines

### PR DESCRIPTION
**Pull request by benaadams**
* Lazy create EndpointRouting statemachines
* Lazy create EndpointMiddleware statemachine

image

Normally complete sync; so skip the async machines in the common path

/cc @rynowak @pranavkm

### CODE REVIEW

File: src/Http/Routing/src/EndpointMiddleware.cs [hunk 1]
```diff
@@ -58,17 +58,34 @@ public async Task Invoke(HttpContext httpContext)
...
+                    var requestTask = endpoint.RequestDelegate(httpContext);
+                    if (!requestTask.IsCompletedSuccessfully)
+                    {
+                        return AwaitRequestTask(endpoint, requestTask, _logger);
+                    }
                 }
-                finally
+                catch (Exception exception)
                 {
                     Log.ExecutedEndpoint(_logger, endpoint);
```
Comment (campersau) [inline src/Http/Routing/src/EndpointMiddleware.cs]: What if the `requestTask` was completed synchronously, then this wouldn't get logged right?

**Comment by pranavkm** [comment]
For my own benefit, how much of an impact does this have? I was planning on looking at the AuthZ middleware at some point and it would be nice to know this makes a reasonable difference before I dive in

File: src/Http/Routing/src/EndpointMiddleware.cs (hunk 1 above)
Comment (benaadams) [reply src/Http/Routing/src/EndpointMiddleware.cs]: Good catch

**Comment by benaadams** [comment]
> For my own benefit, how much of an impact does this have on throughput?

Depends how many; ResourceInvoker was quite good (+3%) https://github.com/aspnet/AspNetCore/pull/9311 but that removed 6; so maybe +0.5% per item on a `dotnet new mvc` + the [`PlainTextActionResult`](https://github.com/aspnet/Benchmarks/blob/6a91ebd2dc574e0ed1f13a4a76930b51944825b3/src/Benchmarks/Controllers/HomeController.cs#L42-L55)

Would be less if the controller was doing more; so it had less impact on Json for example

File: src/Http/Routing/src/EndpointMiddleware.cs (hunk 1 above)
Comment (benaadams) [reply src/Http/Routing/src/EndpointMiddleware.cs]: Fixed

File: src/Http/Routing/src/EndpointRoutingMiddleware.cs [hunk 2]
```diff
@@ -53,15 +53,24 @@ internal sealed class EndpointRoutingMiddleware
...
+            }
+
+            var matchTask = matcherTask.Result.MatchAsync(httpContext, feature);
+            if (!matchTask.IsCompletedSuccessfully)
+            {
+                return AwaitMatch(this, httpContext, feature, matchTask);
+            }
 
-            await matcher.MatchAsync(httpContext, feature);
             if (feature.Endpoint != null)
```
Comment (Tratcher) [inline src/Http/Routing/src/EndpointRoutingMiddleware.cs]: Can you call CompleteAwaited here to avoid the code duplication? (And give ti a better name?)

Comment (benaadams) [reply src/Http/Routing/src/EndpointRoutingMiddleware.cs]: Sure

**Review by rynowak — APPROVED** [review]

**Pull request merged**